R/BackendHdf5-class.R

```diff
@@ -0,0 +1,225 @@
+            " not found."
+        ))
+    }
+    NULL
+}
+
+setValidity("BackendHdf5", function(object) {
+    msg <- c(
+        .valid.BackendHdf5.h5files(object@files, object@h5files),
+        .valid.BackendHdf5.checksums(object@checksums, object@h5files)
+    )
+    if (is.null(msg)) { TRUE } else { msg }
+})
+
+#' Initialize the Hdf5 backend, i.e. create the hdf5 files and add their names
+#' to the @hdf5file slot.
+#'
+#' @param object `BackendHdf5`
+#'
+#' @param files `character` with the original file names.
+#'
+#' @param spectraData `DataFrame` with the spectrum metadata.
+#'
+#' @param path `character(1)` defining the path where the hdf5 files should be
+#'     stored.
+#'
+#' @return `BackendHdf5` object.
+#'
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @md
+#'
+#' @noRd
+setMethod(
+    "backendInitialize",
+    "BackendHdf5",
+    function(object, files, spectraData, path=".", ...) {
+    path <- normalizePath(path, mustWork=FALSE)
+    dir.create(path, showWarnings=FALSE, recursive=TRUE)
+
+    files <- normalizePath(files)
+    object@h5files <- file.path(path, paste0(.vdigest(files), ".h5"))
+
+    if (any(file.exists(object@h5files))) {
+        stop(
+            "File(s) ",
+            paste0(object@h5files[file.exists(object@h5files)], collapse=", "),
+            " already exist(s). Please choose a different 'path'."
+        )
+    }
+
+    for (i in seq(along=object@h5files)) {
+        h5 <- H5Fcreate(object@h5files[i])
+        h5createGroup(h5, "spectra")
+        H5Fclose(h5)
+    }
+
+    object@checksums <- .vdigest(object@h5files, file=TRUE)
+
+    callNextMethod()
+})
+
+#' Import the data from the raw MS files and store them to the hdf5 files.
+#'
+#' @inheritParams backendInitialize
+#'
+#' @return `BackendHdf5`
+#'
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @md
+#'
+#' @noRd
+setMethod(
+    "backendImportData",
+    "BackendHdf5",
+    function(object, spectraData, ..., BPPARAM=bpparam()) {
+
+    bpmapply(object@files, object@h5files, FUN=.mzr2hdf5, BPPARAM=BPPARAM)
+    object@checksums <- .vdigest(object@h5files, file=TRUE)
+    validObject(object)
+    object
+})
+
+setMethod(
+    "backendReadSpectra",
+    "BackendHdf5",
+    function(object, spectraData, ...) {
+
+    uFileIdx <- unique(spectraData$fileIdx)
+    h5files <- object@h5files[uFileIdx]
+
+    changed <- .valid.BackendHdf5.checksums(object@checksums[uFileIdx], h5files)
+    if (!is.null(changed))
+        stop(changed)
+
+    spectra <- vector(mode="list", length=nrow(spectraData))
+    spectra <- split(spectra, spectraData$fileIdx)
+    spd <- split(spectraData, spectraData$fileIdx)
+    spIdx <- split(spectraData$spIdx, spectraData$fileIdx)
+
+    for (i in seq(along=h5files)) {
+        h5fh <- H5Fopen(h5files[i])
+        h5gfh <- H5Gopen(h5fh, "spectra")
+        spectra[[i]] <- .spectra_from_data(
+            lapply(as.character(spIdx[[i]]), h5read, file=h5gfh), spd[[i]]
```

Comment:
Honestly, I wouldn't use `rhdf5::h5read`. While being the *official* exported function to read data from an Hdf5 file it is also slow if multiple small matrices (like in our case) are read (because of the many validations that are performed).

I spent a lot of time figuring out how to improve the speed of the backend in the *hdf5_jo* branch. We should use code from there (for which also unit tests exist).

Reply:
Sorry, but I wasn't familiar with all your experiments/investigations you did before. But the speed gain you achieved is really impressive (10-15x on my laptop :+1::+1:)